### PR TITLE
feat: add plugin manager decorator for MSCA

### PR DESCRIPTION
**Pull request by moldy530**
# Pull Request Checklist

- [ ] Did you add new tests and confirm existing tests pass? (`yarn test`)
- [ ] Did you update relevant docs? (docs are found in the `site` folder, see guidleines below)
- [ ] Do your commits follow the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) standard?
- [ ] Does your PR title also follow the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) standard?
- [ ] If you have a breaking change, is it [correctly reflected in your commit message](https://www.conventionalcommits.org/en/v1.0.0/#examples? (e.g. `feat!: breaking change`)
- [ ] Did you run lint (`yarn lint:check`) and fix any issues? (`yarn lint:fix`)
- [ ] Is the base branch you're merging into `development` and not `main`?

---

## PR-Codex overview
This PR focuses on improving the plugin management system in the codebase. 

### Detailed summary:
- Added support for the `generate` script in `nx.json`.
- Added a new `generate` script in `package.json` which runs the `generate` command.
- Updated the `build` script in `package.json` to include a `lint:write` command after the build process.
- Added new types and imports related to plugin management.
- Updated the `useAlchemyProvider` hook in `aa-simple-dapp` to use the `createMultiOwnerMSCA` function instead of `LightSmartContractAccount`.
- Added new files and changes related to plugin management in the `accounts` package.
- Added new files and changes related to plugin management in the `core` package.
- Updated the `provider` package to include new types and imports related to plugin management.
- Updated the `SmartAccountProvider` class in the `provider` package to include support for provider decorators.
- Updated the `disconnect` method in the `SmartAccountProvider` class to include support for provider decorators.
- Added new files and changes related to plugin management in the `msca` package.
- Added new exports and types related to plugin management in the `accounts` package.

> The following files were skipped due to too many changes: `packages/core/src/account/types.ts`, `packages/accounts/src/msca/plugins/multi-owner.ts`, `packages/accounts/scripts/plugingen.ts`, `packages/accounts/src/msca/abis/IPluginManager.ts`, `packages/accounts/src/msca/builder.ts`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

**Review by avasisht23 — COMMENTED** [review]
some questions on setup, thanks for porting this over from 6900

File: packages/accounts/src/msca/abis/IPluginManager.ts [hunk 1]
```diff
@@ -0,0 +1,223 @@
+export const IPluginManagerAbi = [
```
[the remark this answers is not in the document]
Comment (moldy530) [reply packages/accounts/src/msca/abis/IPluginManager.ts:1]: I think we should do that as part of the productionization process. These things aren't final yet (they most likely are) so we should wait for audit to complete

File: packages/accounts/src/msca/plugins/manager/decorator.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import type { ISmartAccountProvider } from "@alchemy/aa-core";
+import type { MSCA } from "../../builder";
+import { installPlugin, type InstallPluginParams } from "./installPlugin.js";
+import {
+  uninstallPlugin,
+  type UninstallPluginParams,
+} from "./uninstallPlugin.js";
+
+export const pluginManagerDecorator = <
```
[the remark this answers is not in the document]
Comment (moldy530) [reply packages/accounts/src/msca/plugins/manager/decorator.ts:9]: yea a decorator is the `@` thing usually, but this is also meant to be used with the Decorator Pattern defined on provider and account: https://refactoring.guru/design-patterns/decorator

File: packages/accounts/src/msca/plugins/manager/installPlugin.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+      })
+      .then((manifest) => {
+        return keccak256(
+          encodeFunctionResult({
+            abi: IPluginAbi,
+            functionName: "pluginManifest",
+            result: manifest,
+          })
+        );
+      }));
```
[the remark this answers is not in the document]
Comment (moldy530) [reply packages/accounts/src/msca/plugins/manager/installPlugin.ts:42]: ummm I was being lazy and didn't want a `let` with `if-else` hahah

File: packages/accounts/src/msca/plugins/manager/installPlugin.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    args: [
+      params.pluginAddress,
+      manifestHash,
+      params.pluginInitData ?? "0x",
+      params.dependencies ?? [],
+      params.injectedHooks ?? [],
+    ],
+  });
+
+  return provider.sendUserOperation(callData);
```
[the remark this answers is not in the document]
Comment (moldy530) [reply packages/accounts/src/msca/plugins/manager/installPlugin.ts:56]: this is the raw calldata to be made to the SCA. the `target, data, value` approach is used to generate call data specific to the defined execution function on the SCA. but this is an escape hatch for when you want to call the SCA at a different method

File: packages/accounts/src/msca/plugins/manager/uninstallPlugin.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    functionName: "uninstallPlugin",
+    args: [
+      params.pluginAddress,
+      params.config ?? "0x",
+      params.pluginUninstallData ?? "0x",
+      params.hookUnapplyData ?? [],
+    ],
+  });
+
+  return provider.sendUserOperation(callData);
```
[the remark this answers is not in the document]
Comment (moldy530) [reply packages/accounts/src/msca/plugins/manager/uninstallPlugin.ts:27]: see above

File: packages/accounts/src/msca/plugins/manager/decorator.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (denniswon) [reply packages/accounts/src/msca/plugins/manager/decorator.ts:9]: @moldy530 how should we integrate with the provider in aa-sdk?
should we have the clients of aa-sdk to directly use this decorator on the SmartAccountProvider (AlchemyProvider or other) on the client side?

[the remark this answers is not in the document]
Comment (moldy530) [reply packages/accounts/src/msca/plugins/manager/decorator.ts:9]: yea the idea is that any `ISmartAccountProvider` can call `.extend(pluginManagerDecorator)` as a result we don't really need to  provide much more than this. I think we can take all of the decorators that we build for MSCA in one big `mscaDecorators` or something to make this easier

File: packages/accounts/src/msca/plugins/manager/installPlugin.ts [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+  const callData = encodeFunctionData({
+    abi: IPluginManagerAbi,
+    functionName: "installPlugin",
+    args: [
+      params.pluginAddress,
+      manifestHash,
+      params.pluginInitData ?? "0x",
+      params.dependencies ?? [],
+      params.injectedHooks ?? [],
```
[the remark this answers is not in the document]
Comment (moldy530) [reply packages/accounts/src/msca/plugins/manager/installPlugin.ts:52]: ah yea good call on the `dependencies` these need to be pulled from the plugin manifest and should default to that if not passed in

[the remark this answers is not in the document]
Comment (moldy530) [reply packages/accounts/src/msca/plugins/manager/installPlugin.ts:52]: I'm not sure about `injectedHooks` lemme double check

[the remark this answers is not in the document]
Comment (moldy530) [reply packages/accounts/src/msca/plugins/manager/installPlugin.ts:52]: ah, the manifest specifies the interfaceIds, so we can't use the plugin manifest to auto populate this field. This field has to be passed on by the user directly

File: packages/accounts/src/msca/plugins/manager/decorator.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (moldy530) [reply packages/accounts/src/msca/plugins/manager/decorator.ts:9]: @denniswon see: https://github.com/alchemyplatform/aa-sdk/pull/287

**Review by avasisht23 — COMMENTED** [review]
last few recs, then lgtm

File: packages/accounts/src/msca/plugins/manager/decorator.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (moldy530) [reply packages/accounts/src/msca/plugins/manager/decorator.ts:9]: Nope, I think it's fine for this to be standalone as it is here. the MSCA builder will just leverage this as is

File: packages/accounts/src/msca/plugins/manager/decorator.ts
[the remark this answers is not in the document]
Comment (moldy530) [reply packages/accounts/src/msca/plugins/manager/decorator.ts:1]: ah yea I'm fine with that change too

**Review by denniswon — APPROVED** [review]

**Review by avasisht23 — APPROVED** [review]
lgtm!

File: packages/accounts/src/msca/plugin-manager/installPlugin.ts [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+import { IPluginManagerAbi } from "../abis/IPluginManager.js";
+import type { MSCA } from "../builder.js";
+import type { InjectedHook } from "./types";
+
+export type InstallPluginParams = {
+  pluginAddress: Address;
+  manifestHash?: Hash;
+  pluginInitData?: Hash;
+  dependencies?: Address[];
+  injectedHooks?: InjectedHook[];
```
[the remark this answers is not in the document]
Comment (avasisht23) [reply packages/accounts/src/msca/plugin-manager/installPlugin.ts:19]: this is on the interface: https://eips.ethereum.org/EIPS/eip-6900#ipluginmanagersol

**Pull request merged**